localstack/services/awslambda/lambda_executors.py

```diff
@@ -199,10 +219,15 @@ def execute(
         version=None,
         asynchronous=False,
         callback=None,
+        lock_discriminator: str = None,
     ):
         def do_execute(*args):
             @cloudwatched("lambda")
             def _run(func_arn=None):
+                lock = None
+                if lock_discriminator:
+                    lock = LAMBDA_ASYNC_LOCKS.locks[lock_discriminator]
+                    lock.acquire()
```

Comment:
I'd say let's try to avoid calling `acquire()`/`release()` directly - there is still the vague chance that the cleanup doesn't work properly (e.g., exception in `finally` block before we reach `release()`). 

Let's better use `with` blocks for any locking related logic:
```
with lock:
    # critical section ...
```

If needed, we can wrap the code in another inline function, to avoid that the nesting level becomes too deep.

Reply:
nullcontext() is a python 3.7 function, but localstack should still support python 3.6.
A 3.6 compatible solution would be using contextlib.ExitStack like this:
```
            with contextlib.ExitStack() as stack:
                    if lock_discriminator:
                        stack.enter_context(LAMBDA_ASYNC_LOCKS.locks[lock_discriminator])
                    // regular logic
```
I guess the question is, should we honor 3.6 support here?

Reply:
Good catch. We're also using nullcontext in `edge.py` (via the `empty_context_manager()` util function) but I'm fine with `contextlib.ExitStack()` - looks like a good solution here 👍